### ISSUE TITLE
Questions releated to the cross attention ?

### ISSUE THREAD

**Issue by batman47steam** [MoaazK/comp511-project#1]
Thank you for sharing the code. I'd like to ask a question regarding the MHSA module. In the original text, the output description seems to suggest attention weights, but why does the code directly output the result of V multiplied by attention?
![image](https://github.com/MoaazK/unet-transformer/assets/36287035/a29e0c96-8e32-4140-ad14-b56cf491c675)

**Comment by MoaazK** [comment]
Hi, I could not exactly understand what you are asking. The MHSA module that I implemented uses MultiheadAttention module made available by PyTorch. The MultiheadAttention module calculates the attention as was originally mentioned by Vaswani et al. (2017) in their "Attention is All You Need" paper. Following is the formula of Attention calcualtion:

$$
\text{Attention}(Q, K, V) = \text{softmax}\left(\frac{QK^T}{\sqrt{d_k}}\right)V
$$

This formula is the output of MHSA module (after multiplying with **V**). Please let me know if you did not understand.

**Comment by batman47steam** [comment]
Hello, thank you for sharing the code and your prompt response. At first, when I read "The computed weight," I thought it might refer to the `attention weights` before the multiplication with the `value` tensor in the function `attn_output, attn_output_weights = multihead_attn(query, key, value)`, something like `_, attn_weight = self.mha(y, y, s, need_weights=True)` in the code. However, I now understand that if it is `attn_weight`, its shape cannot be (h, w, d), but rather something like (hw * hw).
https://github.com/MoaazK/unet-transformer/blob/3150bd4c15cb5ad547580b6086d8ca44166f516c/src/models/TransformerUNet.py#L99
![image](https://github.com/MoaazK/unet-transformer/assets/36287035/12b4b6c8-f171-420a-a9d5-4d92a45f09ff)
